src/nvim/tui/input.c

```diff
@@ -558,16 +559,107 @@ HandleState ut_handle_background_color(TermInput *input)
 }
 #endif
 
+// During startup we query the terminal to see if it supports the kitty keyboard
+// protocol. If the terminal responds with only the primary device attributes
+// then it does not support the kitty keyboard protocol; if it first responds
+// with the current progressive enhancement flags then it does.
+//
+// See https://sw.kovidgoyal.net/kitty/keyboard-protocol/#detection-of-support-for-this-protocol
+static HandleState handle_extkeys(TermInput *input)
+{
+  if (input->waiting_for_extkeys_response <= 0) {
+    return kNotApplicable;
+  }
+
+  size_t buf_size = rbuffer_size(input->read_stream.buffer);
+  if (buf_size >= 5
+      && !rbuffer_cmp(input->read_stream.buffer, "\x1b[?64", 5)) {
+    return kComplete;
+  } else if (buf_size >= 6
+             && !rbuffer_cmp(input->read_stream.buffer, "\x1b[=", 3)) {
+
+  } else {
+    input->waiting_for_extkeys_response--;
+    if (input->waiting_for_extkeys_response == 0) {
+      DLOG("did not get a response for terminal extended keys support");
+    }
+    return kNotApplicable;
+  }
+
+  bool bad = false;
+  bool done = false;
+  size_t count = 0;
+  int flags = 0;
+  int mode = 1;
+  RBUFFER_EACH(input->read_stream.buffer, c, i) {
+    count = i + 1;
+
+    // Skip the first 3 bytes (CSI =)
+    if (i < 3) {
+      continue;
+    }
+
+    if (i == 3) {
+      flags = c;
+    } else if (c == ';') {
+      continue;
+    } else if (c == 'u') {
+      done = true;
+      break;
+    } else if (c >= 1 && c <= 3) {
+      mode = c;
+    } else {
+      bad = true;
+    }
+  }
+
+  // Now we need to read the primary device attribute response
+  RBUFFER_EACH(input->read_stream.buffer, c, i) {
+    count = i + 1;
+
+    // Skip the first 3 bytes (CSI ?)
+    if (i < 3) {
+      continue;
+    }
+
+    // Read until we encounter a 'c'
+    if (c == 'c') {
+      done = true;
+      break;
+    }
+  }
+
+  if (done && !bad) {
+    rbuffer_consumed(input->read_stream.buffer, count);
+    DLOG("extkeys response: flags=%d, mode=%d", flags, mode);
+    // 0b1 = Disambiguate escape codes (what we call "extended keys")
+    bool supports_kitty_protocol = flags & 0x1;
+    loop_schedule_deferred(&main_loop,
+                           event_create(tui_enable_extkeys, 2, supports_kitty_protocol, input));
```

Comment:
Do not call `tui_enable_extkeys` in the main thread.

Reply:
Which event loop should it run on then?

Reply:
Unless you meant it shouldn't be scheduled on a loop at all and I should just call it directly?

Reply:
Would it have been easier to say that `loop_schedule_deferred` does not need to be used? The processing of `tui_enable_extkeys` must be done in the `TUI` thread. It is fine to refer to existing code in your implementation, but you must understand what that code is doing.